arm cross compile failed at target tbb on master while ok on 2.4

##### System information (version)
- OpenCV => master commit 0d86681b of may 30
- Operating System / Platform => ubuntu 16.04 64-bit cross for arm
- Compiler => arm-linux-gnueabihf-gcc-4.7/g++-4.7

##### Detailed description
Cross-compilation for arm failed on master while ok on 2.4 branch. 

- Commands

```
cmake -DENABLE_VFPV3=ON -DENABLE_NEON=ON \
 -DCMAKE_CXX_COMPILER=/usr/bin/arm-linux-gnueabihf-g++-4.7 \
 -DCMAKE_C_COMPILER=/usr/bin/arm-linux-gnueabihf-gcc-4.7 \
 -DWITH_TBB=ON -DBUILD_TBB=ON \
 -DCMAKE_TOOLCHAIN_FILE=../arm-gnueabi.toolchain.cmake ../../..

make
```

- Error messages

```
[ 36%] Building CXX object 3rdparty/tbb/CMakeFiles/tbb.dir/tbb-4.4.3/src/rml/client/rml_tbb.cpp.o
[ 36%] Building CXX object 3rdparty/tbb/CMakeFiles/tbb.dir/arm_linux_stub.cpp.o
/home/.../opencv/3rdparty/tbb/arm_linux_stub.cpp:6:6: error: prototype for 'void tbb::internal::affinity_helper::protect_affinity_mask()' does not match any in class 'tbb::internal::affinity_helper'
In file included from /home/.../opencv/3rdparty/tbb/arm_linux_stub.cpp:1:0:
/home/.../opencv/platforms/linux/masterbuild/3rdparty/tbb/tbb-4.4.3/src/tbb/tbb_misc.h:255:14: error: candidate is: void tbb::internal::affinity_helper::protect_affinity_mask(bool)
3rdparty/tbb/CMakeFiles/tbb.dir/build.make:854: recipe for target '3rdparty/tbb/CMakeFiles/tbb.dir/arm_linux_stub.cpp.o' failed
make[2]: *** [3rdparty/tbb/CMakeFiles/tbb.dir/arm_linux_stub.cpp.o] Error 1
CMakeFiles/Makefile2:638: recipe for target '3rdparty/tbb/CMakeFiles/tbb.dir/all' failed
make[1]: *** [3rdparty/tbb/CMakeFiles/tbb.dir/all] Error 2
Makefile:160: recipe for target 'all' failed
make: *** [all] Error 2
```

- The failing head

```
$ git log -1
commit 0d86681b88d4605b49ea06e0e6218c1bcc36925c
Merge: fdba548 0f42031
Author: Alexander Alekhin <alexander.a.alekhin@gmail.com>
Date:   Tue May 30 20:11:33 2017 +0000

    Merge pull request #8823 from wpalfi:patch-1
```

##### Steps to reproduce

Install dependency packages on a ubuntu x64 host. 

Following the instruction from http://docs.opencv.org/2.4/doc/tutorials/introduction/crosscompilation/arm_crosscompile_with_cmake.html , clone opencv from github, then build. It will fail. 

Checkout 2.4 branch, and build. Everything is ok.

Tried the patch from #8844 it passed but failed later at linking opencv_test_core. 

* Command to configure and build, note that now it builds static libraries: 
```
cd ~/opencv/opencv
mkdir build
cd build
cmake -DENABLE_VFPV3=ON -DENABLE_NEON=ON \
 -DCMAKE_CXX_COMPILER=/usr/bin/arm-linux-gnueabihf-g++-4.7 \
 -DCMAKE_C_COMPILER=/usr/bin/arm-linux-gnueabihf-gcc-4.7 \
 -DWITH_TBB=ON -DBUILD_TBB=ON \
 -DBUILD_SHARED_LIBS=OFF \
 -DCMAKE_TOOLCHAIN_FILE=../platforms/linux/arm-gnueabi.toolchain.cmake ..
```

* Error message: 
```
[ 55%] Building CXX object modules/core/CMakeFiles/opencv_test_core.dir/test/test_conjugate_gradient.cpp.o
[ 55%] Building CXX object modules/core/CMakeFiles/opencv_test_core.dir/test/test_intrin.cpp.o
[ 55%] Linking CXX executable ../../bin/opencv_test_core
../../3rdparty/lib/libtbb.a(governor.cpp.o): In function `tbb::internal::governor::release_resources()':
governor.cpp:(.text._ZN3tbb8internal8governor17release_resourcesEv+0xe): undefined reference to `tbb::internal::destroy_process_mask()'
collect2: error: ld returned 1 exit status
modules/core/CMakeFiles/opencv_test_core.dir/build.make:871: recipe for target 'bin/opencv_test_core' failed
make[2]: *** [bin/opencv_test_core] Error 1
CMakeFiles/Makefile2:1711: recipe for target 'modules/core/CMakeFiles/opencv_test_core.dir/all' failed
make[1]: *** [modules/core/CMakeFiles/opencv_test_core.dir/all] Error 2
Makefile:160: recipe for target 'all' failed
make: *** [all] Error 2
```

Ok, it still fails with a clean build or with toolchain 4.9. Here is an exact sequence to reproduce it if you have docker.io installed on the ubuntu 16.04.2 host. 

* Create a directory and place the following in a file named `bb.txt` in the directory. 
```
FROM ubuntu:xenial-20170214
RUN apt-get update
RUN apt-get install -y python git make vim ninja-build
RUN apt-get install -y gcc g++ cmake
RUN apt-get install -y gcc-4.7-arm-linux-gnueabihf
RUN apt-get install -y g++-4.7-arm-linux-gnueabihf
```
* Go into the directory and create the docker image named `testimg`, then start it as a container named `testcont`, and start a shell in the container:
```
docker build -f bb.txt -t testimg .
docker run -dt --name testcont testimg /bin/bash
docker exec -it testcont /bin/bash
```
* Inside the container, clone opencv:
```
# cd
# mkdir opencv
# cd opencv
# git clone https://github.com/opencv/opencv
```
* Patch the tbb file as in #8844
* Build
```
# mkdir build
# cd build
# cmake -GNinja \
-DCMAKE_TOOLCHAIN_FILE=../opencv/platforms/linux/arm-gnueabi.toolchain.cmake \
-DCMAKE_CXX_COMPILER=/usr/bin/arm-linux-gnueabihf-g++-4.7 \
-DCMAKE_C_COMPILER=/usr/bin/arm-linux-gnueabihf-gcc-4.7 \
-DWITH_TBB=ON -DBUILD_TBB=ON \
-DBUILD_SHARED_LIBS=OFF \
-DENABLE_NEON=ON -DENABLE_VFPV3=ON \
../opencv
```
* The same problem will happen, but sometimes on `tbb_version`, or on `opencv_annotation`. 
```
[682/1076] Linking CXX executable bin/opencv_annotation
[683/1076] Linking CXX executable bin/opencv_version
FAILED: : && /usr/bin/arm-linux-gnueabihf-g++-4.9   -mthumb  -fdata-sections -Wa,--noexecstack -fsigned-char -Wno-psabi -fPIC   -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wno-narrowing -Wno-delete-non-virtual-dtor -Wno-comment -fdiagnostics-show-option -pthread -fomit-frame-pointer -ffunction-sections  -mfpu=vfpv3 -mfpu=neon -mfp16-format=ieee -mfpu=neon -fvisibility=hidden -fvisibility-inlines-hidden -O3 -DNDEBUG  -DNDEBUG  -Wl,--fix-cortex-a8 -Wl,--no-undefined -Wl,--gc-sections -Wl,-z,noexecstack -Wl,-z,relro -Wl,-z,now  -Wl,-z,nocopyreloc apps/version/CMakeFiles/opencv_version.dir/opencv_version.cpp.o  -o bin/opencv_version  -rdynamic -lstdc++ -ldl -lm -lpthread -lrt 3rdparty/lib/libtbb.a lib/libopencv_core.a lib/libopencv_highgui.a lib/libopencv_imgproc.a lib/libopencv_imgcodecs.a lib/libopencv_videoio.a lib/libopencv_imgcodecs.a lib/libopencv_imgproc.a lib/libopencv_core.a -lstdc++ -lpthread -lrt 3rdparty/lib/libtbb.a -lc -ldl -lm 3rdparty/lib/liblibjpeg.a 3rdparty/lib/liblibwebp.a 3rdparty/lib/liblibpng.a 3rdparty/lib/liblibtiff.a 3rdparty/lib/liblibjasper.a 3rdparty/lib/libIlmImf.a 3rdparty/lib/libzlib.a 3rdparty/lib/libtegra_hal.a && :
3rdparty/lib/libtbb.a(governor.cpp.o): In function `tbb::internal::governor::release_resources()':
governor.cpp:(.text._ZN3tbb8internal8governor17release_resourcesEv+0xe): undefined reference to `tbb::internal::destroy_process_mask()'
collect2: error: ld returned 1 exit status
FAILED: : && /usr/bin/arm-linux-gnueabihf-g++-4.9   -mthumb  -fdata-sections -Wa,--noexecstack -fsigned-char -Wno-psabi -fPIC   -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wno-narrowing -Wno-delete-non-virtual-dtor -Wno-comment -fdiagnostics-show-option -pthread -fomit-frame-pointer -ffunction-sections  -mfpu=vfpv3 -mfpu=neon -mfp16-format=ieee -mfpu=neon -fvisibility=hidden -fvisibility-inlines-hidden -O3 -DNDEBUG  -DNDEBUG  -Wl,--fix-cortex-a8 -Wl,--no-undefined -Wl,--gc-sections -Wl,-z,noexecstack -Wl,-z,relro -Wl,-z,now  -Wl,-z,nocopyreloc apps/annotation/CMakeFiles/opencv_annotation.dir/opencv_annotation.cpp.o  -o bin/opencv_annotation  -rdynamic -lstdc++ -ldl -lm -lpthread -lrt 3rdparty/lib/libtbb.a lib/libopencv_core.a lib/libopencv_highgui.a lib/libopencv_imgproc.a lib/libopencv_imgcodecs.a lib/libopencv_videoio.a lib/libopencv_imgcodecs.a lib/libopencv_imgproc.a lib/libopencv_core.a -lstdc++ -lpthread -lrt 3rdparty/lib/libtbb.a -lc -ldl -lm 3rdparty/lib/liblibjpeg.a 3rdparty/lib/liblibwebp.a 3rdparty/lib/liblibpng.a 3rdparty/lib/liblibtiff.a 3rdparty/lib/liblibjasper.a 3rdparty/lib/libIlmImf.a 3rdparty/lib/libzlib.a 3rdparty/lib/libtegra_hal.a && :
3rdparty/lib/libtbb.a(governor.cpp.o): In function `tbb::internal::governor::release_resources()':
governor.cpp:(.text._ZN3tbb8internal8governor17release_resourcesEv+0xe): undefined reference to `tbb::internal::destroy_process_mask()'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

--- a/3rdparty/tbb/arm_linux_stub.cpp
+++ b/3rdparty/tbb/arm_linux_stub.cpp
@@ -3,8 +3,9 @@
 namespace tbb {
 namespace internal {
 
-void affinity_helper::protect_affinity_mask() {}
+void affinity_helper::protect_affinity_mask(bool) {}
 affinity_helper::~affinity_helper() {}
+void destroy_process_mask() {}
 
 }
 }

